common/djangoapps/entitlements/tasks.py

```diff
@@ -62,3 +72,49 @@ def expire_old_entitlements(self, start, end, logid='...'):
         raise self.retry(exc=exc, countdown=countdown, max_retries=MAX_RETRIES)
 
     LOGGER.info('Successfully completed the task expire_old_entitlements after examining %d entries [%s]', entitlements.count(), logid)  # lint-amnesty, pylint: disable=line-too-long
+
+
+@shared_task(bind=True, ignore_result=True)
+@set_code_owner_attribute
+def expire_and_create_entitlements(self, no_of_entitlements):
+    """
+    This task is designed to be called to process and expire bundle of entitlements
+    that are older than one year on in exceptional case 18 months.
+
+    Args:
+        None
+
+    Returns:
+        None
+
+    """
+    LOGGER.info('Running task expire_and_create_entitlements')
+
+    current_date = date.today()
+    expiration_period = current_date - relativedelta(years=1)
+    exceptional_expiration_period = current_date - relativedelta(years=1, months=6)
+    normal_entitlements = CourseEntitlement.objects.filter(
+        expired_at__isnull=True, created__lte=expiration_period).exclude(course_uuid__in=MIT_SUPPLY_CHAIN_COURSES)
+    exceptional_entitlements = CourseEntitlement.objects.filter(
+        expired_at__isnull=True, created__lte=exceptional_expiration_period, course_uuid__in=MIT_SUPPLY_CHAIN_COURSES)
+
+    entitlements = normal_entitlements | exceptional_entitlements
+    countdown = 2 ** self.request.retries
+
+    try:
+        for entitlement in entitlements[:no_of_entitlements]:
+
+            entitlement.expire_entitlement()
+            LOGGER.info('Expired entitlement with id %d ', entitlement.id)
+            entitlement.pk = None
```

Comment:
Is this the right way to do this -- modifying the existing `entitlement` instead of creating a new var?

This also does not looks quite like the way the support tool currently does it, here: https://github.com/openedx/frontend-app-support-tools/blob/b9a73c54d6aea895dc148dd97d9bef08b246c575/src/users/entitlements/CreateEntitlementForm.jsx#LL31C1-L42C1.

In particular, `refund_locked` is not set in this creation.

Reply:
Ok, so this may work, but I still have some concerns about how clear it is what will happen.  This approach hides what is staying the same, instead of explicitly setting each property.

For example, when creating a new entitlement using support-tools, `refund_locked` is explicitly set to `true`.  But here it is unclear whether `refund_locked` will be `true` or `false`, because it will depend on whatever the previous value was.

Reply:
Looks like `CourseEntitlement.uuid` is a unique field:

https://github.com/openedx/edx-platform/blob/41c27e8cbde0c215ef867bf5af6b679ece29d040/common/djangoapps/entitlements/models.py#L162

@zubair-ce07, how about we set `entitlement.uuid = None` so it'll default to a new UUID?

Reply:
1. Is there a reason not to just create a new CourseEntitlement instead of trying to reuse `entitlement`?  This would avoid any errors like leaving `uuid` the same.
2. The intention of this new entitlement is to be a $0 grant to the user.  It should be exactly like creating a new entitlement using support-tools.  So it should be correct to set `refund-locked` to `True`.
3. Now that I look at this, we will also end up keeping the order number, which is _not_ going to work correctly.  We need that to be `None` so that the financial reporting will not associate this new entitlement with money.  We also need to not have a course run set.  Even more reason to please just create a new CourseEntitlement and explicitly set fields, instead of keeping fields from the old one.  We should set it exactly like [support-tools does](https://github.com/openedx/frontend-app-support-tools/blob/f39774824c6b232aaba50b2dd725c01d529b81cd/src/users/entitlements/CreateEntitlementForm.jsx#LL31C1-L41C9).  This is the only stuff that should be set:
```    postEntitlement({
      requestData: {
        course_uuid: courseUuid,
        user,
        mode,
        refund_locked: true,
        support_details: [{
          action: CREATE,
          comments,
        }],
      },```

Reply:
Gotcha, I was misunderstanding the requirements. That approach makes sense to me.

Reply:
Done! Creating a new object of CourseEntitlement and setting only those fields that support tool does.